### PR TITLE
RA-968 Built module for OpenMRS 2.0 platform

### PR DESCRIPTION
**Pull request by jdegraft**

### CODE REVIEW

File: api/pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 	<parent>
 		<groupId>org.openmrs.module</groupId>
 		<artifactId>appframework</artifactId>
-		<version>2.5-SNAPSHOT</version>
+		<version>2.6-SNAPSHOT</version>
```
Comment (wluyima) [inline api/pom.xml]: Why are you changing this?

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java [hunk 2]
```diff
@@ -66,12 +71,46 @@
...
+	private User saveUser(UserService us, User u, String password)
+	{
+		User user = null;
+		Class<?> userServiceClass = us.getClass();
+    	try {
+    		// openmrs 1.9
+    		Method saveUserMethod =  userServiceClass.getDeclaredMethod("saveUser", u.getClass(), password.getClass());
+    		user = (User)saveUserMethod.invoke(us, u, password);
+    		return user;
+    	} catch (NoSuchMethodException ex) {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java]: I know we normally do this but i think it's cleaner to just catch ReflectiveOperationException since it's a common super class

File: api/pom.xml (hunk 1 above)
Comment (wluyima) [inline api/pom.xml]: Is this 2.5 already released?

File: omod/pom.xml [hunk 3]
```diff
@@ -67,6 +79,13 @@
 			<groupId>org.openmrs.module</groupId>
 			<artifactId>webservices.rest-omod-common</artifactId>
 		</dependency>
+       	
+		<dependency>
+	        <groupId>org.javassist</groupId>
```
Comment (wluyima) [inline omod/pom.xml]: Do both versions work with this version of javassist? I would assume you would define different javassist dependencies inside the specific profiles

File: omod/pom.xml [hunk 4]
```diff
@@ -196,4 +215,64 @@
...
+				<dependencies>	
+					<dependency>
+						<groupId>org.openmrs.module</groupId>
+						<artifactId>legacyui-omod</artifactId>
+						<version>${legacyUIVersion}</version>
+						<type>jar</type>
+						<scope>provided</scope>
+					</dependency>
+					<dependency>
+						<groupId>org.openmrs.module</groupId>
```
Comment (wluyima) [inline omod/pom.xml]: Seeme like this is a common dependency with the same version and scope, why not take it out of the profiles?

File: omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java [hunk 5]
```diff
@@ -1,100 +0,0 @@
...
-import org.openmrs.module.appframework.service.AppFrameworkService;
-import org.openmrs.module.appframework.domain.AppDescriptor;
-import org.openmrs.util.PrivilegeConstants;
-import org.openmrs.web.controller.PortletController;
-import org.springframework.web.servlet.ModelAndView;
-
-/**
- *
- */
-public class AppFrameworkHomepagePortletController extends PortletController {
```
Comment (wluyima) [inline omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java]: Why did you get rid of this? It's still needed for 1.9

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java [hunk 6]
```diff
@@ -139,14 +136,16 @@ public void testGetAllExtensionsAndIsSortedByOrder() throws Exception {
...
+        List<Extension> extensionPoints = appFrameworkService.getAllExtensions("extensionPoint2");
 
         assertEquals(2, extensionPoints.size());
         assertEquals("ext5", extensionPoints.get(0).getId());
-        assertEquals("ext2", extensionPoints.get(1).getId());
+        assertEquals("ext4", extensionPoints.get(1).getId());
     }
 
     @Test
+    @Ignore("can't get this to work")
```
Comment (djazayeri) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: we should look into this before merging

File: omod/pom.xml [hunk 7]
```diff
@@ -67,6 +79,13 @@
 			<groupId>org.openmrs.module</groupId>
 			<artifactId>webservices.rest-omod-common</artifactId>
 		</dependency>
+       	
+		<dependency>
+	        <groupId>org.javassist</groupId>
+	        <artifactId>javassist</artifactId>
+	        <version>3.18.2-GA</version>
+	    </dependency>
```
Comment (djazayeri) [inline omod/pom.xml]: shouldn't this have scope=test?

File: omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java (hunk 5 above)
Comment (djazayeri) [inline omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java]: I said to get rid of this, actually. We can drop support for the legacy UI within this module with this upcoming release, because nobody is actually using that feature.

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 8]
```diff
@@ -26,14 +26,16 @@
     	 <property name="order"><value>98</value></property> 
     	<property name="mappings">
             <props>
+            <!--
```
Comment (djazayeri) [inline omod/src/main/resources/webModuleApplicationContext.xml]: don't comment this out, remove it

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 9]
```diff
@@ -26,14 +26,16 @@
...
     	<property name="mappings">
             <props>
+            <!-- 
 				<prop key="**/appFrameworkHomepage.portlet">appFrameworkHomepagePortletController</prop>
+             -->
 				<prop key="**/module/appframework">appFrameworkController</prop>
 			</props>
 		</property>
 	</bean>
+	<!--
```
Comment (djazayeri) [inline omod/src/main/resources/webModuleApplicationContext.xml]: don't comment this out, remove it

File: pom.xml [hunk 10]
```diff
@@ -93,6 +94,14 @@
 				<version>${webservicesRestVersion}</version>
 				<scope>provided</scope>
 			</dependency>
+			
+			<dependency>
+				<groupId>org.openmrs.module</groupId>
+				<artifactId>legacyui-omod</artifactId>
```
Comment (djazayeri) [inline pom.xml]: Why do we need this? I think that App Framework should be dropping support for the Legacy UI.

File: api/pom.xml (hunk 1 above)
Comment (jdegraft) [inline api/pom.xml]: @darius, wanted a major new release, because the changes removed support for 1.x line

File: omod/pom.xml (hunk 3 above)
Comment (jdegraft) [inline omod/pom.xml]: yes, I have built all profiles/versions and they run fine with this version of javaassist

File: omod/pom.xml (hunk 4 above)
Comment (jdegraft) [inline omod/pom.xml]: profile 1.9 does not use this module that is why.

File: api/pom.xml (hunk 1 above)
Comment (wluyima) [inline api/pom.xml]: But 2.5 is meant to be the next major release

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java [hunk 11]
```diff
@@ -53,23 +53,26 @@
...
+    	allAppDescriptors.clear();
+
         featureToggles.setPropertiesFile(new File(this.getClass().getResource("/" + FeatureToggleProperties.FEATURE_TOGGLE_PROPERTIES_FILE_NAME).getFile()));
 
         app1 = new AppDescriptor("app1", "desc1", "label1", "url1", "iconUrl", "tinyIconUrl", 20,
                 null, Arrays.asList(new ExtensionPoint("extensionPoint1"), new ExtensionPoint("extensionPoint2")));
 
         app2 = new AppDescriptor("app2", "desc1", "label1", "url1", "iconUrl", "tinyIconUrl", 10,
-                null, Arrays.asList(new ExtensionPoint("extensionPoint1")));
+                null, Arrays.asList(new ExtensionPoint("extensionPoint2")));
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: Why did you change from  extensionPoint1 to 2?

File: omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java (hunk 5 above)
Comment (wluyima) [inline omod/src/main/java/org/openmrs/module/appframework/web1x/AppFrameworkHomepagePortletController.java]: I see

File: api/pom.xml (hunk 1 above)
Comment (wluyima) [inline api/pom.xml]: Sorry not 2.5 but 3.0 would be the next major release

Comment (wluyima) [inline api/pom.xml]: Did you see this comment @jdegraft ?

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java [hunk 12]
```diff
@@ -53,23 +53,26 @@
 
     private Extension ext5;
 
-
     @Before
     public void setUp() throws Exception {
 
+    	// clear up from previous tests
```
Comment (djazayeri) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: Why did you move these here instead of leaving them in @After? The other way was much more standard.

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 11 above)
Comment (djazayeri) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: broadly speaking, why do you have to change the things this test is actually testing?

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java [hunk 13]
```diff
@@ -66,12 +71,42 @@
...
+    		// openmrs 1.9
+    		Method saveUserMethod =  userServiceClass.getDeclaredMethod("saveUser", u.getClass(), password.getClass());
+    		user = (User)saveUserMethod.invoke(us, u, password);
+    		return user;
+    	} catch (ReflectiveOperationException ex) {
+    		// do nothing
+		}
+    	try {
+    		// openmrs 2.0
+    		Method saveUserMethod =  userServiceClass.getDeclaredMethod("saveUser", u.getClass());
```
Comment (djazayeri) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java]: why get this method if you never call it?

File: omod/pom.xml [hunk 14]
```diff
@@ -196,4 +216,56 @@
 		</plugins>
 	</build>
 
+	<!-- different profiles for building/testing against different version of
```
Comment (djazayeri) [inline omod/pom.xml]: Do we need to duplicate the profiles in this omod/pom.xml in addition to the root pom.xml?

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 12 above)
Comment (jdegraft) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: The current tests use DirtiesContext and the semantics have changed in the spring/hibernate versions used in platform 2.0. I had to remove the DirtiesContext annotation to get the tests to run and had to make changes in the appframework tests to compensate.

After DirtiesContext is invoked, it is impossible to login into the platform hence tests fail.

File: pom.xml [hunk 15]
```diff
@@ -215,4 +227,54 @@
...
+	<!-- different profiles for building/testing against different version of 
+		openmrs -->
+	<profiles>
+		<profile>
+			<id>openmrs-1.9</id>
+
+			<properties>
+				<openMRSVersion>1.9.9</openMRSVersion>
+				<openMRSMinorVersion>1.9</openMRSMinorVersion>
+				<webservicesRestVersion>2.13-SNAPSHOT</webservicesRestVersion>
```
Comment (djazayeri) [inline pom.xml]: Why would we test against different versions of the REST module? I would prefer that we just make this module require the latest (now snapshot) version of REST, and make that common in both profiles...

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java (hunk 13 above)
Comment (jdegraft) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java]: It was a way to ensure that we are using 2.0 api.

File: pom.xml (hunk 15 above)
Comment (jdegraft) [inline pom.xml]: I could do that. Daniel seemed to prefer a 1.x profile that was identical.

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java (hunk 13 above)
Comment (djazayeri) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceTest.java]: If that's the case, then say so in the comment.

File: pom.xml (hunk 15 above)
Comment (djazayeri) [inline pom.xml]: @dkayiwa ?

Comment (jdegraft) [inline pom.xml]: yes, @dkayiwa

Comment (dkayiwa) [inline pom.xml]: @jdegraft am fine with what @djazayeri suggested.

**Comment by jdegraft** [comment]
@dkayiwa, there is still the outstanding issue with DirtiesContext. I tried to remove it however this undermined the integrity of the current tests. Personally I am not happy with changing tests and also there is still one test which does not run.
We need to decide how to handle this otherwise this change cannot really be merged.

**Comment by dkayiwa** [comment]
@jdegraft can you give us a summary of how to reproduce that problem?

**Comment by jdegraft** [comment]
Well all you have to do is clone the module and build against platform 2.0. You will have to make changes to support java 8 which includes changing the javaassist dependency.
For the spring/hibernate versions used in platform 2.0, when DirtiesContext causes the applicationContext to be refreshed, the database is flushed. This causes the admin user to be removed hence the tests which require authentication fail. In fact because no user exists, it is not possible to authenticate manually to pass the tests.
appframework tests make changes to the applicationContext, so without DirtiesContext, it is not possible for the tests to run with a clean slate.

**Comment by dkayiwa** [comment]
@jdegraft this is not clear to me. :)
Can you just create a pull request to reproduce it? With only the minimal changes.

**Comment by jdegraft** [comment]
@dkayiwa I did this several weeks ago and you rejected the pull request because there were too many changes.
You have built modules for platform 2.0 several times yourself, (you guided me on the sort of the changes I should be making for RA-954 in general), I think I should pause work on this PR, considering how important appframework is, and what I think the DirtiesContext problem is, as DirtiesContext is used in other modules, until you have time to look into it.
Consider this a blocker for RA-954 in general. :)

**Comment by dkayiwa** [comment]
@jdegraft does this mean that if i do not make all the changes you made in the pull request, then i cannot reproduce it? :)

**Comment by jdegraft** [comment]
@dkayiwa, this pull request includes more than necessary to run the tests. It includes additional changes such as removing support for web 1.x. We should shelf this pull request for now.
To see the DirtiesContext problem all that is necessary is to build the original module against platform 2.0 which only involves changes to the pom, particularly the javaassist dependencies.
What I am suggesting is you start from afresh and just build the original module for platform 2.0 without the additional changes just to evaluate the DirtiesContext problem.
Once we know how we are going to handle the DirtiesContext issue, if you find that it is a problem, I can redo this pull request. If you don't find a problem then you can suggest how I can redo the pull request.:)

**Comment by dkayiwa** [comment]
@jdegraft i tried building the module against platform 2.0 and never got the dirties context problem. That is the basis of my asking for a pull request with the minimal changes to reproduce it, in case there is something i missed out. :)

**Comment by jdegraft** [comment]
@dkayiwa good point. This raises the issue of whether I am approaching RA-954 correctly.
My understanding is when building for platform 2.0, the 2.0 pom profile must use openmrs-api-2.0.jar, etc whiles the 1.9 pom profile must use openmrs-api-1.9.jar etc.
Without using core 2.0 jars, you will not be using the new spring/hibernate versions and the DirtiesContext problem will not be visible.
This is how the example pom profiles @wluyima  created were structured.
Let me know if this is indeed correct.

**Comment by dkayiwa** [comment]
I think i will understand better with a pull request. :)

**Comment by jdegraft** [comment]
@dkayiwa, is it possible to extract the pom files from the pull request and use those? There is a 2.0 profile setup.

File: pom.xml [hunk 16]
```diff
@@ -106,8 +106,20 @@
...
+	                    <artifactId>javassist</artifactId>
+	                </exclusion>
+	            </exclusions>
         	</dependency>
-
+        	
+			<dependency>
+		        <groupId>org.javassist</groupId>
+		        <artifactId>javassist</artifactId>
+		        <version>3.18.2-GA</version>
```
Comment (wluyima) [inline pom.xml]: I thought the scope needs to be test for this

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 12 above)
Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: The reason DirtiesContext was added was to remove the apps and extension added by the test after it's done, so Darius is right you need to call these in a method annotated with @After, and i guess with that you won't need to change the tests

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 11 above)
Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: See the comment aboveusing @After so that you don't have to change these tests

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 6 above)
Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: @jdegraft are you looking into fixing this?

File: omod/pom.xml (hunk 14 above)
Comment (wluyima) [inline omod/pom.xml]: @jdegraft did you see this comment?

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 12 above)
Comment (jdegraft) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: @Wyclif, DirtiesContext in openmrs 2.0 causes the database to be emptied. Tests are not able to proceed because authentication is not possible. I agree we need DirtiesContext to maintain the tests in appframework but we need to make changes to the platform to support DirtiesContext. Particularly, I proposed a change where if Authentication in core failed, the database was reloaded.
I think a more senior dev has to solve the problem with DirtiesContext before this ticket can be resolved.
It is possible that DirtiesContext is not a problem, but this really needs to be verified by a senior dev because I am seeing a problem.

Comment (wluyima) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: I'm not sure if it's necessary to make a change in core, the only reason we has @DirtiesContext was to have the apps cleaned up as the application context gets reloaded, since you got rid of @DirtiesContext annotation, a manual clean up of the apps seems like the obvious solution which you are doing, all we are saying is that it should happen after each test and not before and my guess is that this has nothing to do with authentication unless may be there is something else you're doing.

**Comment by dkayiwa** [comment]
@jdegraft i have made a couple of commits to the module to make it compile fine on 2.x without changing the core platform. So am closing this pull request. If you need to do any followup commit, feel free to create a new pull request. Otherwise thank you so much for exploring this. Keep up the great work!!! :+1:

File: api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java (hunk 12 above)
Comment (wyclif) [inline api/src/test/java/org/openmrs/module/appframework/service/AppFrameworkServiceImplTest.java]: Please stop sending me your pull requests. I'm getting copied on them by
mistake. Thanks.
